pageserver/src/walredo.rs

```diff
@@ -587,419 +583,1287 @@ impl<C: CommandExt> CloseFileDescriptors for C {
+                        while let Ok(permit) = feedback.downsize.try_acquire() {
+                            permit.forget();
+                            forgotten_permits += 1;
+                        }
+
+                        if forgotten_permits > 0 {
+                            warn!(forgotten_permits, "this should really have a test case for it, but we just forgot permits while adding a new process");
+                        }
+
+                        true
+                    },
+                    _ = feedback.notify_not_empty.notified(), if check_not_empty => {
+                        successive_empty = 0;
+                        continue;
+                    },
+                    _ = feedback.notify_empty.notified(), if js.len() > 1 => {
+                        downsized_at_limit |= js.len() >= max_processes.get();
+                        // let one exit
+                        // TODO: this lets many since all of them will notify at same time ...
+                        // could debounce to some tick rate?
+                        feedback.downsize.add_permits(1);
+                        successive_empty = 0;
+                        continue;
+                    },
+                    _ = feedback.notify_empty.notified(), if js.len() == 1 => {
+                        successive_empty += 1;
+                        if successive_empty > 15 {
+                            // if we have been idle for 15 timeouts, then we might as well kill the
+                            // last process and go to awaiting for the next work in this task.
+                            feedback.downsize.add_permits(1);
+                        }
+                        continue;
+                    },
+                    exited = js.join_next() => {
+                        match exited {
+                            Some(Ok(())) => {
+                                // normal, logging happens in walredo_rpc
+                                continue;
+                            },
+                            Some(Err(e)) => {
+                                // hook should had printed already
+                                warn!("walredo task panicked: {e}");
+                                continue;
+                            },
+                            None => {
+                                if cancel.is_cancelled() {
+                                    // on shutdown, now the children have all exited, so should we.
+                                    break;
+                                } else if first.is_some() {
+                                    // we are yet to launch our first process
+                                    true
+                                } else {
+                                    debug!("all wal redo processes have been stopped, waiting for the next redo request before launching new process");
+                                    continue;
```

Comment:
```suggestion
                                    break;
```

This fixes the hanging. Logic bug which prevented the scale back up from zero.

Reply:
Made tests for this, tucked them nicely within scale to zero and scale up. That was an interesting exercise.